lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php

```diff
@@ -299,50 +302,67 @@ public function executeInserts()
     /**
      * Retrieves the default version value which was created
      * by the preceding INSERT statement and assigns it back in to the
-     * entities version field.
+     * entities version field if the given entity is versioned.
+     * Also retrieves values of columns marked as 'non insertable' and / or
+     * 'not updatetable' and assigns them back to the entities corresponding fields.
      *
      * @param object  $entity
      * @param mixed[] $id
      *
      * @return void
      */
-    protected function assignDefaultVersionValue($entity, array $id)
+    protected function assignDefaultVersionAndUpsertableValues($entity, array $id)
     {
-        $value = $this->fetchVersionValue($this->class, $id);
+        $values = $this->fetchVersionAndNotUpsertableValues($this->class, $id);
 
-        $this->class->setFieldValue($entity, $this->class->versionField, $value);
+        foreach ($values as $field => $value) {
+            $value = Type::getType($this->class->fieldMappings[$field]['type'])->convertToPHPValue($value, $this->platform);
+
+            $this->class->setFieldValue($entity, $field, $value);
+        }
     }
 
     /**
-     * Fetches the current version value of a versioned entity.
+     * Fetches the current version value of a versioned entity and / or the values of fields
+     * marked as 'not insertable' and / or 'not updateable'.
      *
      * @param ClassMetadata $versionedClass
      * @param mixed[]       $id
      *
      * @return mixed
      */
-    protected function fetchVersionValue($versionedClass, array $id)
+    protected function fetchVersionAndNotUpsertableValues($versionedClass, array $id)
     {
-        $versionField = $versionedClass->versionField;
-        $fieldMapping = $versionedClass->fieldMappings[$versionField];
-        $tableName    = $this->quoteStrategy->getTableName($versionedClass, $this->platform);
-        $identifier   = $this->quoteStrategy->getIdentifierColumnNames($versionedClass, $this->platform);
-        $columnName   = $this->quoteStrategy->getColumnName($versionField, $versionedClass, $this->platform);
+        $columnNames = [];
+        foreach ($this->class->fieldMappings as $key => $column) {
+            if (isset($column['notInsertable']) || isset($column['notUpdateable']) || ($this->class->isVersioned && $key === $versionedClass->versionField)) {
+                $columnNames[$key] = $this->quoteStrategy->getColumnName($key, $versionedClass, $this->platform);
+            }
+        }
+
+        $tableName  = $this->quoteStrategy->getTableName($versionedClass, $this->platform);
+        $identifier = $this->quoteStrategy->getIdentifierColumnNames($versionedClass, $this->platform);
 
         // FIXME: Order with composite keys might not be correct
-        $sql = 'SELECT ' . $columnName
-             . ' FROM ' . $tableName
-             . ' WHERE ' . implode(' = ? AND ', $identifier) . ' = ?';
+        $sql = 'SELECT ' . implode(', ', $columnNames)
```

Comment:
You need to use `$column as $fieldName` here otherwise the following access of `$this->class->fieldMappings[$field]` based on `$values` will not work when column name != field name.